flink-connectors/flink-connector-cassandra/src/test/java/org/apache/flink/batch/connectors/cassandra/CassandraOutputFormatBaseTest.java

```diff
@@ -0,0 +1,334 @@
+        }
+    }
+
+    @Test
+    public void testReleaseOnSuccess() throws Exception {
+        try (TestCassandraOutputFormat openedTestCassandraOutputFormat =
+                createOpenedTestCassandraOutputFormat()) {
+            assertThat(openedTestCassandraOutputFormat.getAvailablePermits()).isEqualTo(1);
+            assertThat(openedTestCassandraOutputFormat.getAcquiredPermits()).isEqualTo(0);
+
+            CompletableFuture<ResultSet> completableFuture = new CompletableFuture<>();
+            openedTestCassandraOutputFormat.enqueueCompletableFuture(completableFuture);
+            openedTestCassandraOutputFormat.writeRecord("N/A");
+
+            assertThat(openedTestCassandraOutputFormat.getAvailablePermits()).isEqualTo(0);
+            assertThat(openedTestCassandraOutputFormat.getAcquiredPermits()).isEqualTo(1);
+
+            // start writing
+            completableFuture.complete(null);
+
+            assertThat(openedTestCassandraOutputFormat.getAvailablePermits()).isEqualTo(1);
+            assertThat(openedTestCassandraOutputFormat.getAcquiredPermits()).isEqualTo(0);
+        }
+    }
+
+    @Test
+    public void testReleaseOnFailure() throws Exception {
+        TestCassandraOutputFormat testCassandraOutputFormat =
+                createOpenedTestCassandraOutputFormat();
+        assertThat(testCassandraOutputFormat.getAvailablePermits()).isEqualTo(1);
+        assertThat(testCassandraOutputFormat.getAcquiredPermits()).isEqualTo(0);
+
+        CompletableFuture<ResultSet> completableFuture = new CompletableFuture<>();
+        testCassandraOutputFormat.enqueueCompletableFuture(completableFuture);
+        testCassandraOutputFormat.writeRecord("N/A");
+
+        assertThat(testCassandraOutputFormat.getAvailablePermits()).isEqualTo(0);
+        assertThat(testCassandraOutputFormat.getAcquiredPermits()).isEqualTo(1);
+
+        completableFuture.completeExceptionally(new RuntimeException());
+
+        assertThat(testCassandraOutputFormat.getAvailablePermits()).isEqualTo(1);
+        assertThat(testCassandraOutputFormat.getAcquiredPermits()).isEqualTo(0);
+        try {
+            testCassandraOutputFormat.close();
+        } catch (IOException ignored) {
+            // the aim is not to assert on the exception in this test
+        }
+    }
+
+    @Test
+    public void testReleaseOnThrowingSend() throws Exception {
+        Function<String, ListenableFuture<ResultSet>> failingSendFunction =
+                ignoredMessage -> {
+                    throw new RuntimeException("expected");
+                };
+
+        try (TestCassandraOutputFormat testCassandraOutputFormat =
+                createOpenedMockOutputFormat(failingSendFunction)) {
+            assertThat(testCassandraOutputFormat.getAvailablePermits()).isEqualTo(1);
+            assertThat(testCassandraOutputFormat.getAcquiredPermits()).isEqualTo(0);
+
+            assertThatThrownBy(() -> testCassandraOutputFormat.writeRecord("none"));
```

Comment:
I think I'd prefer an explicit isNotNull clause or something along those lines; it threw me off that this actually asserts anything (because that wouldn't be the case for assertThat())